torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -374,6 +374,361 @@ def collect_bw_donated_buffer_idxs(
+            elif node.op == "output":
+                old_outputs = node.args[0]
+                new_outputs = (
+                    *old_outputs[-num_fw_outputs:],
+                    *old_outputs[:-num_fw_outputs],
+                )
+                new_outputs = [env[n] if n else None for n in new_outputs]
+                new_graph.output(tuple(new_outputs))
+            else:
+                env[node] = new_graph.node_copy(node, lambda n: env[n])
+                env[node].meta = copy.copy(node.meta)
+
+        new_graph.lint()
+
+        out = torch.fx.GraphModule(joint_gm, new_graph)
+        return out
+
+    new_hop_graphs: Dict[str, InvokeSubgraphHopGraphs] = defaultdict(
+        lambda: InvokeSubgraphHopGraphs()
+    )
+
+    # Step 1 - Get a `joint_hop_gm` from the `fw_hop_gm` and `bw_hop_gm` This is
+    # easy to do for `invoke_subgraph` HOP. During the Autograd dispatch key
+    # tracing, we have put the joint_hop_graph in the backward hop graph itself.
+    # So to recover the joint_hop_gm, we just have to look at the backward
+    # HOP graphs.
+    # So we will merge step 1 and step 2 in this next section
+
+    for node in joint_gm.graph.nodes:
+        if (
+            node.op == "call_function"
+            and node.target is invoke_subgraph
+            and isinstance(node.args[1], str)
+        ):
+            identifier = (
+                node.args[1].replace("___forward", "").replace("___backward", "")
+            )
+
+            # If partitioning already done for this identifier, skip.
+            if new_hop_graphs[identifier].partitioning_done:
+                continue
+
+            hop_gm = getattr(joint_gm, node.args[0].target)
+            assert isinstance(hop_gm, torch.fx.GraphModule)
+
+            if node.args[1].startswith("___forward"):
+                # Collect some information from the forward hop graph
+                new_hop_graphs[identifier].old_num_fw_inputs = num_inputs(hop_gm)
+                new_hop_graphs[identifier].old_num_fw_outputs = num_outputs(hop_gm)
+            elif node.args[1].startswith("___backward"):
+                num_fw_inputs = new_hop_graphs[identifier].old_num_fw_inputs
+                assert num_fw_inputs is not None
+                num_fw_outputs = new_hop_graphs[identifier].old_num_fw_outputs
+                assert num_fw_outputs is not None
+
+                # Step 1) - Get the `joint_hop_gm`. As mentioned earlier, the
+                # backward graph is the joint graph.
+                joint_hop_gm = hop_gm
+
+                # Prepare the graph for the partitioner
+                joint_hop_gm = prepare_for_partitioner(
+                    joint_hop_gm, num_fw_inputs, num_fw_outputs
+                )
+
+                # Step 2) and 3) - Run joint graph passes and partitioner
+                new_fw_hop_gm, new_bw_hop_gm = aot_config.partition_fn(
+                    joint_hop_gm, [], num_fwd_outputs=num_fw_outputs
+                )
```

Comment:
Does this actually run joint_graph_passes? https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/joint_graph.py#L521

Reply:
It does. `test_dropout` is written to check exactly that.